rtos/rtx5/mbed_rtx_idle.cpp

```diff
@@ -0,0 +1,198 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "rtos/rtos_idle.h"
+#include "platform/mbed_sleep.h"
+#include "TimerEvent.h"
+#include "us_ticker_api.h"
+#include "core_cm.h"
+#include "mbed_critical.h"
+#include "mbed_assert.h"
+#include <new>
+#include "rtx_os.h"
+extern "C" {
+#include "rtx_lib.h"
+}
+
+using namespace mbed;
+
+#if (defined(NO_SYSTICK))
+extern "C" IRQn_Type mbed_get_m0_tick_irqn(void);
+#endif
+
+class RtosTimer : private TimerEvent {
+public:
+    RtosTimer(): TimerEvent(get_us_ticker_data()), start_time(0), tick(0) {
+        start_time = ticker_read_us(_ticker_data);
+#if (defined(NO_SYSTICK))
+        NVIC_SetVector(mbed_get_m0_tick_irqn(), (uint32_t)SysTick_Handler);
+        NVIC_SetPriority(mbed_get_m0_tick_irqn(), 0xFF); /* RTOS requires lowest priority */
+        NVIC_EnableIRQ(mbed_get_m0_tick_irqn());
+#else
+        // Ensure SysTick has the correct priority as it is still used
+        // to trigger software interrupts on each tick. The period does
+        // not matter since it will never start counting.
+        SysTick_Setup(16);
+#endif
+    };
+
+    /**
+     * Schedule an os tick to fire
+     *
+     * @param delta Tick to fire at relative to current tick
+     */
+    void schedule_tick(uint32_t delta=1) {
+        insert_absolute(start_time + (tick + delta) * 1000);
+    }
+
+
+    /**
+     * Prevent any scheduled ticks from triggering
+     */
+    void cancel_tick() {
+        remove();
+    }
+
+    /**
+     * Get the current tick count
+     *
+     * @return The number of ticks since boot. This should match RTX's tick count
+     */
+    uint32_t get_tick() {
+        return tick & 0xFFFFFFFF;
+    }
+
+    /**
+     * Update the internal tick count
+     *
+     * @return The number of ticks incremented
+     */
+    uint32_t update_tick() {
+        uint64_t new_tick = ticker_read_us(_ticker_data) / 1000;
+        if (new_tick > tick) {
+            // Don't update to the current tick.
+            // The RTOS must do this to properly
+            // schedule tasks again.
+            new_tick--;
+        }
+        uint32_t elapsed_ticks = new_tick - tick;
+        tick = new_tick;
+        return elapsed_ticks;
+    }
+
+    /**
+     * Get the time
+     *
+     * @return Current tmie in microseconds
+     */
+    us_timestamp_t get_time() {
+        return ticker_read_us(_ticker_data);
+    }
+
+    ~RtosTimer() {
+
+    };
+
+protected:
+
+    void handler() {
+#if (defined(NO_SYSTICK))
+        NVIC_SetPendingIRQ(mbed_get_m0_tick_irqn());
+#else
+        SCB->ICSR = SCB_ICSR_PENDSTSET_Msk;
+#endif
+        tick++;
+        schedule_tick();
+    }
+
+    us_timestamp_t start_time;
+    uint64_t tick;
+};
+
+static RtosTimer *os_timer;
+static uint64_t os_timer_data[sizeof(RtosTimer) / 8];
+
+/// Setup System Timer.
+int32_t osRtxSysTimerSetup (void)
```

Comment:
Isn't there CMSIS-RTOS for tick API? What I found is https://github.com/ARM-software/CMSIS_5/blob/6e5621b3bc4becf50ac5f29a07f30ff2081a5e70/CMSIS/RTOS2/Include/os_tick.h , but that seems to be not in our codebase - https://www.keil.com/pack/doc/CMSIS/RTOS2/html/group__CMSIS__RTOS__TickAPI.html. I guess we get this once we update RTOS

Just to be aware

Reply:
They've changed the names again, the API was there and is used by nrf51_dk https://github.com/ARMmbed/mbed-os/blob/5f138810a90e913f617f4345654709eb11281385/rtos/rtx5/TARGET_CORTEX_M/rtx_system.c#L212